Make OrderBy binary compat across TFMs (#68036)
dotnet/roslyn: docs/Target Framework Strategy.md

Target Framework Strategy
===

# Layers
The roslyn repository produces components for a number of different products that push varying ship and TFM constraints on us. A summary of some of our dependencies are : 

- Build Tools: requires us to ship compilers on `net472`
- .NET SDK: requires us to ship compilers on current servicing target framework (presently `net6.0`)

> - Source build: requires us to `$(NetCurrent)` and `$(NetPrevious)` in workspaces and below (presently `net8.0` and `net7.0` respectively)

Reviewer: ```suggestion
- Source build: requires us to ship `$(NetCurrent)` and `$(NetPrevious)` in workspaces and below (presently `net8.0` and `net7.0` respectively)
```

- Visual Studio: requires us to ship `net472` for base IDE components and `net6.0` for private runtime components.

It is not reasonable for us to take the union of all TFM and multi-target every single project to them. That would add several hundred compilations to any build operation which would in turn negatively impact our developer throughput. Instead we attempt to use the TFM where needed. That keeps our builds smaller but increases complexity a bit as we end up shipping a mix of TFM for binaries across our layers.

# Require consistent API across Target Frameworks
It is important that our shipping APIs maintain consistent API surface area across target frameworks. That is true whether the API is `public` or `internal`.

The reason for `public` is standard design pattern. The reason for `internal` is a combination of the following problems:

> - Our repository make use of `InternalsVisibleTo` which effectively promotes `internal` APIs to `public` ones.

Reviewer 2: Not sure if this change helps, but maybe?

```suggestion
- Our repository make use of `InternalsVisibleTo` which allows other assemblies to directly reference signatures of `internal` members.
```

> - Our repository make use of `InternalsVisibleTo` which allows other assemblies to directly reference signatures of `internal` members.

Reviewer: ```suggestion
- Our repository makes use of `InternalsVisibleTo` which allows other assemblies to directly reference signatures of `internal` members.
```

- Our repository ships a mix of target frameworks. Typically workspaces and below will ship more recent TFMs than the layers above it. Compiler has to ship newer TFM for source build while IDE is constrained by Visual Studio's private runtime hence adopts newer TFM slower.
- Our repository invests in polyfill APIs to make compiling against multiple TFMs in the same project a seamless experience.

> Taken together though this means that our `internal` surface area in many cases effectively `public` when it comes to binary compatibility. For example a consuming project can end up with `net7.0` binaries from workspaces layer and `net6.0` binaries from IDE layer. Because there is `InternalsVisibleTo` between these binaries the `internal` API surface area is effectively `public`. This requires us to have a consistent strategy for achieving binary compatibility across TFM combinations.

Reviewer 3: ```suggestion
Taken together though this means that our `internal` surface area in many cases is effectively `public` when it comes to binary compatibility. For example a consuming project can end up with `net7.0` binaries from workspaces layer and `net6.0` binaries from IDE layer. Because there is `InternalsVisibleTo` between these binaries the `internal` API surface area is effectively `public`. This requires us to have a consistent strategy for achieving binary compatibility across TFM combinations.
```

Consider a specific example of what goes wrong when our `internal` APIs are not consistent across TFM:

> - Workspaces today targets `net6.0` and `net7.0` and it contains our `EnumerableExtensions.cs` which polyfills many extensions methods on `IEnumerable<T>`. In `net7.0` the `OrderBy` extension method is not needed because it was put into the .NET core libraries.

Reviewer 4: Nit: 

```suggestion
- Workspaces today targets `net6.0` and `net7.0` and it contains our `EnumerableExtensions.cs` which polyfills many extensions methods on `IEnumerable<T>`. In `net7.0` the `Order` extension method is not needed because it was put into the .NET core libraries.
```

> - Language Server Protocol targets `net6.0` and consumes the `OrderyBy` polyfill from Workspaces

Reviewer: ```suggestion
- Language Server Protocol targets `net6.0` and consumes the `Order` polyfill from Workspaces
```

> Lets assume for a second that we `#if` the `OrderBy` method such that it's not present in `net7.0`.  Locally this all builds because we compile the `net6.0` versions against each other so they're consistent. However if an external project which targets `net7.0` and consumes both Workspaces and Language Server Protocol then it will be in a broken state. The Protocol binary is expecting Workspaces to contain a polyfill method for `OrderBy` but it does not since it's at `net7.0` and it was `#if` out. As a result this will fail at runtime with missing method exceptions.

Reviewer 2: ```suggestion
Let's assume for a second that we `#if` the `OrderBy` method such that it's not present in `net7.0`.  Locally this all builds because we compile the `net6.0` versions against each other so they're consistent. However if an external project which targets `net7.0` and consumes both Workspaces and Language Server Protocol then it will be in a broken state. The Protocol binary is expecting Workspaces to contain a polyfill method for `OrderBy` but it does not since it's at `net7.0` and it was `#if` out. As a result this will fail at runtime with missing method exceptions.
```

> Let's assume for a second that we `#if` the `OrderBy` method such that it's not present in `net7.0`.  Locally this all builds because we compile the `net6.0` versions against each other so they're consistent. However if an external project which targets `net7.0` and consumes both Workspaces and Language Server Protocol then it will be in a broken state. The Protocol binary is expecting Workspaces to contain a polyfill method for `OrderBy` but it does not since it's at `net7.0` and it was `#if` out. As a result this will fail at runtime with missing method exceptions.

Reviewer: ```suggestion
Let's assume for a second that we `#if` the `Order` method such that it's not present in `net7.0`.  Locally this all builds because we compile the `net6.0` versions against each other so they're consistent. However if an external project which targets `net7.0` and consumes both Workspaces and Language Server Protocol then it will be in a broken state. The Protocol binary is expecting Workspaces to contain a polyfill method for `Order` but it does not since it's at `net7.0` and it was `#if` out. As a result this will fail at runtime with missing method exceptions.
```

This problem primarily comes from our use of polyfill APIs. To avoid this we employ the following rule:

> > When there is an `#if NET.*` that declares a non-private member, there must be an `#else` that defines an equivalent binary compatible symbol

Reviewer 2: In this case, `#if NET.* is a regular expression, but it could be confused as a glob by the reader where `.` is literal.

Author: Tried to clean that up a bit.

Reviewer 5: Why not use `NET... ` everywhere for consistency?

> This comes up in two forms 

Reviewer 2: ```suggestion
This comes up in two forms:
```

## Pattern for types 

> When creating a polyfill for a type use the `#if !NET...` to declare the type and in the `#else` use a `TypeForwardedTo` for the actual type.

Reviewer 5: ❓ Why `!` here? Generally that's harder to read if we always have an `else`

Author: I agree it's harder to read and I considered saying we do the other approach. However most of the type polyfills were written using `#if !NET...` so I was giving way to precedence here. If there is concensus to change I'm happy to put in the keyboard time to fix it up.

Author: @chsienki, @sharwell thoughts?

Reviewer 4: Given the small-ish number of them, I vote we flip it and make it easier to read.

Reviewer 3: Flip em.

Example: 

```csharp
#if !NET6_0_OR_GREATER

using System.ComponentModel;

namespace System.Runtime.CompilerServices
{
    internal static class IsExternalInit
    {
    }
}

#else

using System.Runtime.CompilerServices;

#pragma warning disable RS0016 // Add public types and members to the declared API (this is a supporting forwarder for an internal polyfill API)
[assembly: TypeForwardedTo(typeof(IsExternalInit))]
#pragma warning restore RS0016 // Add public types and members to the declared API

#endif
```

## Pattern for extension methods
When creating a polyfill for an extension use the `#if NET...` to declare the extension method and the `#else` to declare the same method without `this`. That will put a method with the expected signature in the binary but avoids it appearing as an extension method within that target framework.

```csharp
#if NET7_0_OR_GREATER

>         public static IOrderedEnumerable<T> Order<T>(IEnumerable<T> source) where T : IComparable<T>

Reviewer 3: Feels like there's less of a chance of error if we just `#if` the `this` keyword, rather than the whole signature.

Author: I thought that too but it looked really awful when I wrote it that way.

Reviewer 5: ```csharp
public static IOrderedEnumerable<T> Order<T>(
    #if !NET7_0_OR_GREATER
    this
    #endif 
    IEnumerable<T> source) where T : IComparable<T>
```

Yea... idk if I like that
